src/transformers/models/vit/modeling_vit.py

```diff
@@ -461,6 +451,18 @@ def _init_weights(self, module: Union[nn.Linear, nn.Conv2d, nn.LayerNorm]) -> No
         elif isinstance(module, nn.LayerNorm):
             module.bias.data.zero_()
             module.weight.data.fill_(1.0)
+        elif isinstance(module, ViTEmbeddings):
+            nn.init.trunc_normal_(
+                module.position_embeddings,
+                mean=0.0,
+                std=self.config.initializer_range,
+            )
+
+            nn.init.trunc_normal_(
+                module.cls_token,
+                mean=0.0,
+                std=self.config.initializer_range,
+            )
```

Comment:
@younesbelkada do you mind filling me in a bit about this refactor? I would greatly appreciate it.

Reply:
hey @sayakpaul !

I just moved the lines above here, before this refactor the attributes `position_embeddings` & `cls_token` were initialized on the fly during their initialization, i.e. whenever we create an instance of ViTEmbedding.
 
But this approach is not what we want to follow since it is preferable to centralize all the weight initialization process inside the method `_init_weights`, for example to call it only when we need it! Consider this snippet:
```
from transformers import ViTModel

model = ViTModel.from_pretrained("google/vit-base-patch16-224-in21k")
```
I think what we want ideally here is to:
1- Create an instance of ViTModel
2- Load the pre-trained weights, straight to the goal 
There is no need to fill the values of the weights with an intendent distribution here, since the weights will be filled at the end by the pre-trained weights. Also note that not calling `_init_weights` will speedup loading large models (this is I believe one of the main reason why it is disabled by default on `.from_pretrained` method).

Therefore that is why we always prefer to do it in two stages:
1- initialize each module and submodules
2- fill the weights of these modules with the correct distribution 
In rare cases you can face unexpected behaviours when doing everything in 1- : e.g. before this PR if you load a ViT model with & `torch_dtype=torch.float16` you'll face an error: 
```
RuntimeError: "erfinv_vml_cpu" not implemented for 'Half'
```
That you can reproduce with this snippet:
```
import torch
import torch.nn as nn

torch.set_default_dtype(torch.float16)

nn.init.trunc_normal_(
    torch.zeros(1, 1, 2),
    mean=0.0,
    std=0.1,
)
```
This is simply because `torch.set_default_dtype(torch.float16)` is called when adding `torch_dtype=torch.float16`, and such errors can be very confusing for users!

Reply:
Beauty! Thanks for being so generous with your explanation!

Reply:
@younesbelkada thank you for tracking the issue